image/cas/layout/tar.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package layout
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/opencontainers/image-spec/image/cas"
+	"github.com/opencontainers/image-spec/image/layout"
+	"golang.org/x/net/context"
+)
+
+// TarEngine is a cas.Engine backed by a tar file.
+type TarEngine struct {
+	file ReadWriteSeekCloser
+}
+
+// NewTarEngine returns a new TarEngine.
+func NewTarEngine(ctx context.Context, file ReadWriteSeekCloser) (eng cas.Engine, err error) {
+	engine := &TarEngine{
+		file: file,
+	}
+
+	err = layout.CheckTarVersion(ctx, engine.file)
+	if err != nil {
+		return nil, err
+	}
+
+	return engine, nil
+}
+
+// Put adds a new blob to the store.
+func (engine *TarEngine) Put(ctx context.Context, reader io.Reader) (digest string, err error) {
+	data, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return "", err
+	}
+
+	size := int64(len(data))
+	hash := sha256.Sum256(data)
```

Comment:
On Tue, Aug 30, 2016 at 03:02:59PM -0700, Stephen Day wrote:
> > +// Put adds a new blob to the store.
> > +func (engine *TarEngine) Put(ctx context.Context, reader io.Reader) (digest string, err error) {
> > +	data, err := ioutil.ReadAll(reader)
> > +	if err != nil {
> > +		return "", err
> > +	}
> > +
> > +	size := int64(len(data))
> > +	hash := sha256.Sum256(data)
> 
> Use the writer version of this, along with `io.TeeReader`.

The reason I don't, as discussed in cda16d26b, is that we aren't
buffering to a temporary disk location and we need to know the hash
before we can *start* writing to the tar file.  And again [1],
tar-write support as this PR stands is just good enough to pass most
unit tests.  I can reroll the tar backend to write to temporary files
if you like [2], but I personally don't see the need to make
tar-writing any more than a toy.  Anybody who has a lot of writing to
do should be using a directory-backed engine or similar which supports
atomic renames.

[1]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76727482
[2]: https://github.com/opencontainers/image-spec/pull/159#issuecomment-243598892